base/compiler/typeinfer.jl

```diff
@@ -235,33 +247,27 @@ function _typeinf(interp::AbstractInterpreter, frame::InferenceState)
               frames[i].cached )
         for i in 1:length(frames) ]
     empty!(frames)
-    if may_optimize(interp)
-        for (caller, _, _) in results
-            opt = caller.src
-            if opt isa OptimizationState
-                result_type = caller.result
-                @assert !(result_type isa LimitedAccuracy)
-                optimize(interp, opt, OptimizationParams(interp), result_type)
-                finish(opt.src, interp)
-                # finish updating the result struct
-                validate_code_in_debug_mode(opt.linfo, opt.src, "optimized")
-                if opt.const_api
-                    if result_type isa Const
-                        caller.src = result_type
-                    else
-                        @assert isconstType(result_type)
-                        caller.src = Const(result_type.parameters[1])
-                    end
-                elseif opt.src.inferred
-                    caller.src = opt.src::CodeInfo # stash a copy of the code (for inlining)
+    for (caller, edges, cached) in results
+        opt = caller.src
+        if may_optimize(interp) && opt isa OptimizationState
+            result_type = caller.result
+            @assert !(result_type isa LimitedAccuracy)
+            optimize(interp, opt, OptimizationParams(interp), result_type)
+            if opt.const_api
+                # XXX: The work in ir_to_codeinf! is essentially wasted. The only reason
+                # we're doing it is so that code_llvm can return the code
+                # for the return (which never runs anyway). We should do this
+                # as a post processing step instead.
+                ir_to_codeinf!(opt)
+                if result_type isa Const
+                    caller.src = result_type
                 else
-                    caller.src = nothing
+                    @assert isconstType(result_type)
+                    caller.src = Const(result_type.parameters[1])
                 end
-                caller.valid_worlds = opt.inlining.et.valid_worlds[]
             end
+            caller.valid_worlds = opt.inlining.et.valid_worlds[]
         end
-    end
-    for (caller, edges, cached) in results
```

Comment:
We do not want to fuse these loops, as it makes inlining unpredictable (refs https://github.com/JuliaLang/julia/pull/38231)

Reply:
Can we just do #38231 first and get inlining predictable then? Otherwise it's much harder to keep the state as IRCode, because inlining might have to deal with seeing something incomplete in an InferenceResult,

Reply:
I guess dealing with that isn't terrible. Inlining already has support for inlining IRCode since Diffractor uses that, so it should be possible to just make it work.

Reply:
To clarify (from #38231), that is indeed the point of having the two separate loops: we want to prevent inlining using these intermediate objects until the second loop has finished marking them as finished. The goal for #38231 is to add a new outer loop around these two that RPO-sorts the results into smaller groups, but the inner strategy here would be the same.